### PR TITLE
Port `BeamSampler` to core

### LINKED ISSUE #1156: Port BeamSampler to multi-backend keras
This was not included in 0.6.0, but we should port it from `tf.xx` -> `ops.xx`.

### PR DESCRIPTION
**Pull request by shivance**
Closes #1156

### CODE REVIEW

**Comment by shivance** [comment]
Hey @mattdangerw !
Could you tell the alternative for these in core?

1. tf.nest.map_structure
2. tf.newaxis
3. tf.while_loop
4. tf.ensure_shape
5. tf.reduce_any
6. tf.reduce_all

**Comment by abheesht17** [comment]
2. ops.expand_dims
3. ops.while_loop
4. You can probably just do x.shape == (…) or something?
5. ops.any
6. ops.all

File: keras_nlp/samplers/beam_sampler.py [hunk 1]
```diff
@@ -186,16 +187,16 @@ def body(prompt, cache, index, log_probs):
 
             def gather_beams(x):
                 x = unflatten_beams(x)
-                x = tf.gather(x, beam_indices, axis=1, batch_dims=1)
+                x = ops.take_along_axis(x, beam_indices, axis=1, batch_dims=1)
```
Comment (mattdangerw) [inline keras_nlp/samplers/beam_sampler.py:190]: `batch_dims` is not a argument to `take_along_axis`

**Comment by mattdangerw** [comment]
For `tf.ensure_shape`, ideally we can ditch it.

**Comment by mattdangerw** [comment]
For `tf.nest.map_structure`, just leave as is for now. On a separate PR, we can update to the `dm_nest` module.

**Comment by shivance** [comment]
> For `tf.ensure_shape`, ideally we can ditch it.

There is this comment at line no. 180 in the file
```
# Ensure shape is `[None]`, otherwise it causes issues after
# converting to TFLite.
```

File: keras_nlp/samplers/beam_sampler.py [hunk 2]
```diff
@@ -213,21 +211,19 @@ def gather_beams(x):
         all_log_probs = unflatten_beams(log_probs)
 
         if self.return_all_beams:
-            sorted_indices = tf.argsort(
-                all_log_probs, axis=-1, direction="DESCENDING"
-            )
-            sorted_log_probs = tf.gather(
-                all_log_probs, sorted_indices, axis=-1, batch_dims=1
+            sorted_indices = ops.argsort(all_log_probs, axis=-1)
```
Comment (shivance) [inline keras_nlp/samplers/beam_sampler.py:214]: We don't have option to set direction in argsort. Because of this couple of tests fail in local with assertion error.

The possible solution could be to reverse the array `sorted_indices` along axis=-1?

Comment (abheesht17) [reply keras_nlp/samplers/beam_sampler.py:214]: ops.argsort(-all_log_probs, axis=-1) should do the trick

Comment (shivance) [reply keras_nlp/samplers/beam_sampler.py:214]: Great! This fixes the failing test

**Comment by mattdangerw** [comment]
> # Ensure shape is `[None]`, otherwise it causes issues after converting to TFLite.

Yeah, I would still ditch and see what happens. This isn't a fix we should need to carry around.

File: keras_nlp/samplers/beam_sampler.py (hunk 1 above)
Comment (shivance) [reply keras_nlp/samplers/beam_sampler.py]: I think this line will throw dimension mismatch between beam_indicea and x

File: keras_nlp/samplers/beam_sampler.py (hunk 2 above)
Comment (abheesht17) [reply keras_nlp/samplers/beam_sampler.py:220]: You need the minus sign in front so that it is sorted in descending order

Comment (shivance) [reply keras_nlp/samplers/beam_sampler.py:220]: Flip does the job here. 

> Great! This fixes the failing test for tf backend

I was referring to ops.flip here 😂. Probably you commented at the same time.

Comment (abheesht17) [reply keras_nlp/samplers/beam_sampler.py:220]: :o. Missed the `ops.flip` line

**Comment by mattdangerw** [comment]
Worked through a lot of this while working through #1187, this ended up being quite tricky. Pushed some fixes.

**Comment by mattdangerw** [comment]
(this last failure is an OOM, unrelated I think)

**Comment by shivance** [comment]
@mattdangerw /gcbrun is needed
